---
title: "qPCR data analysis using R"
output: html_notebook
---

- https://cran.r-project.org/web/packages/pcr/vignettes/qpcr_analysis.html
- https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-7-85
- https://www.sigmaaldrich.com/US/en/technical-documents/technical-article/genomics/qpcr/data-analysis
- https://bitesizebio.com/24894/4-easy-steps-to-analyze-your-qpcr-data-using-double-delta-ct-analysis/
- https://www.nature.com/articles/nprot.2008.73
- http://satqpcr.sophia.inra.fr/
- https://academic.oup.com/clinchem/article/55/4/611/5631762



```{r message=FALSE, warning=FALSE, results='hide'}
library(tidyverse)
library(magrittr)
library(janitor)
library(readxl)
library(pcr)
library(lme4)
library(lmerTest)
```

```{r}
# default mode delta_delta_ct
## locate and read raw ct data
ct1 <- read.csv(system.file('extdata', 'ct1.csv', package = 'pcr'))

## add grouping variable
group_var <- rep(c('brain', 'kidney'), each = 6)

# calculate all values and errors in one step
## mode == 'separate_tube' default
pcr_analyze(ct1,
  group_var = group_var,
  reference_gene = 'GAPDH',
  reference_group = 'brain')

ct1 %>%
  mutate(group = group_var) %>%
  group_by(group) %>%
  summarize(c_myc.mean = mean(c_myc),
            c_myc.sd = sd(c_myc),
            GAPDH.mean = mean(GAPDH),
            GAPDH.sd = sd(GAPDH),
            dct.mean = c_myc.mean - GAPDH.mean,
            dct.sd = sqrt(c_myc.sd^2 + GAPDH.sd^2))
```

```{r}
# default mode delta_delta_ct
## locate and read raw ct data
ct2 = read.csv(system.file('extdata', 'ct2.csv', package = 'pcr'))

## add grouping variable
group_var = rep(c('brain', 'kidney'), each = 6)

# calculate all values and errors in one step
## mode == 'same_tube'
pcr_analyze(ct2,
  group_var = group_var,
  reference_gene = 'GAPDH',
  reference_group = 'brain',
  mode = 'same_tube')

ct2 %>%
  mutate(group = group_var) %>%
  mutate(dct = c_myc - GAPDH) %>%
  group_by(group) %>%
  summarize(dct.mean = mean(dct), dct.sd = sd(dct))
```

```{r}
# locate and read data
ct4 = read.csv(system.file('extdata', 'ct4.csv', package = 'pcr'))

ct4

# make group variable
group_var = rep(c('control', 'treatment'), each = 12)

# analyze the testing data
pcr_analyze(ct4, 
            group_var = group_var,
            reference_gene = 'ref',
            reference_group = 'control')

ct4 %>%
  mutate(group = group_var) %>%
  group_by(group) %>%
  summarize(target.mean = mean(target),
            target.sd = sd(target),
            ref.mean = mean(ref),
            ref.sd = sd(ref),
            dct.mean = target.mean - ref.mean,
            dct.sd = sqrt(target.sd^2 + ref.sd^2))

df = ct4 %>%
  mutate(group = group_var,
         run = factor(rep(c(1:3), 8)),
         dCt = target - ref)

t.test(formula = dCt ~ group, data = df)

summary(lm(dCt ~ group, data = df))

pcr_test(ct4, 
         group_var = group_var,
         reference_gene = 'ref',
         reference_group = 'control',
         test = "lm")

###
df = ct4 %>%
  mutate(group = group_var) %>%
  pivot_longer(c(ref, target), names_to = "gene", values_to = "Ct")

df

summary(lm(Ct ~ group * gene, data = df))
```


```{r}
df = read_xlsx("ct4.xlsx")

df

df1 = df %<>%
  mutate(dCt = Target - Reference)

t.test(dCt ~ Group, data = df1)

wilcox.test(dCt ~ Group, data = df1)

df2 = df %>%
  pivot_longer(c(Target, Reference), names_to = "Gene", values_to = "Ct")

summary(lmer(Ct ~ Gene*Group + Concentration + (1 | Sample), data = df2))
```